patroni/ha.py

```diff
@@ -1402,3 +1415,23 @@ def get_remote_member(self, member=None):
 
     def get_remote_master(self):
         return self.get_remote_member()
+
+    def call_pre_promote(self, config):
```

Comment:
I would really recommend to look at [post_bootstrap](https://github.com/zalando/patroni/blob/master/patroni/ha.py#L1151-L1159) method.

The `pre_promote` script should be executed by a function (or method) with the help of AsyncExecutor. It is hard to get the return code from such method, therefore we are using the [CryticalTask](https://github.com/zalando/patroni/blob/master/patroni/async_executor.py#L7) object. Usage example you can find [here](https://github.com/zalando/patroni/blob/master/patroni/postgresql/bootstrap.py#L336).

If the function executed asynchronously returns not None, the HA loop will be immediately [rescheduled](https://github.com/zalando/patroni/blob/master/patroni/async_executor.py#L107-L108).